tests/python_package_test/test_engine.py

```diff
@@ -938,6 +938,52 @@ def test_early_stopping_via_global_params(first_metric_only):
     assert "error" in gbm.best_score[valid_set_name]
 
 
+@pytest.mark.parametrize("early_stopping_round", [-10, -1, 0, None, "None"])
+def test_early_stopping_non_positive_values(early_stopping_round):
+    X, y = load_breast_cancer(return_X_y=True)
+    num_trees = 5
+    params = {
+        "num_trees": num_trees,
+        "objective": "binary",
+        "metric": "None",
+        "verbose": -1,
+        "early_stopping_round": early_stopping_round,
+        "first_metric_only": True,
+    }
+    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)
+    lgb_train = lgb.Dataset(X_train, y_train)
+    lgb_eval = lgb.Dataset(X_test, y_test, reference=lgb_train)
+    valid_set_name = "valid_set"
+
+    if early_stopping_round is None:
+        gbm = lgb.train(
+            params,
+            lgb_train,
+            feval=[decreasing_metric, constant_metric],
+            valid_sets=lgb_eval,
+            valid_names=valid_set_name,
+        )
+        assert "early_stopping_round" not in gbm.params
```

Comment:
```suggestion
        assert "early_stopping_round" not in gbm.params
        assert gbm.num_trees() == num_trees
```

Testing that the parameter is omitted from the `params` attribute is useful, but I think this test should also be testing that early stopping was not actually triggered.

Reply:
So I actually thought about that, but since the test is set up to be so short, even if early stopping is enabled, the algorithm would hit 5 trees before any type of early stopping would be reasonable to expect, right?  (for example, if I set it to '2', the tree is still build with 5 sub-trees).   So the only way I thought to "ensure" that early stopping didn't happen was to see if the param was dropped.  Am I missing something?

Reply:
> the algorithm would hit 5 trees before any type of early stopping would be reasonable to expect, right

Yep, agreed!

But that expectation should be explicitly tested, so that if some bug causes it to be violated in the future, we find out about it via a test failure. One goal with these tests is to get all these expectations out of humans' heads and into automation, so we don't have to remember all these things every time we review a change.

By passing `"num_trees": 5`, you're telling LightGBM "train for at most 5 rounds".

Confirming that LightGBM actually did train for exactly 5 rounds is the most direct possible test that early stopping wasn't triggered.

Reply:
Looks like you added this in https://github.com/microsoft/LightGBM/pull/6406/commits/260fb9483910f7e52d15365b39448d5adef5ba8c, thanks! I'll resolve this thread.